mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt

```diff
@@ -202,8 +203,29 @@ class WidgetAdapter(
         return false
     }
 
-    private fun getItemViewType(widget: Widget): Int {
+    private fun updateWidgetAtPosition(position: Int, widget: Widget) {
+        val oldWidget = items[position]
+        items[position] = widget
+        widgetsById[widget.id] = widget
+        // If visibility of a container with at least one child changes, refresh the whole list to make sure
+        // the child visibility is also updated. Otherwise it's sufficient to update the single widget only.
+        if (oldWidget.visibility != widget.visibility && items.any { w -> w.parentId == widget.id }) {
+            notifyDataSetChanged()
+        } else {
+            notifyItemChanged(position)
+        }
+    }
+
+    private tailrec fun isWidgetIncludingAllParentsVisible(widget : Widget): Boolean {
```

Comment:
AS claims `tailrec` isn't needed here.

Reply:
Where do you see this? I don't see any notice like that in the code editor, during build or in the 'analyze code' output.

Reply:
This is a lint inspection that is run during editing code.

Reply:
OK. Looks like one would need to help the compiler here:

```diff
diff --git a/mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt b/mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt
index 1a5609c1..ce681cda 100644
--- a/mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt
+++ b/mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt
@@ -220,8 +220,8 @@ class WidgetAdapter(
         if (!widget.visibility) {
             return false
         }
-        val parent = widget.parentId?.let { id -> widgetsById[id] }
-        return parent == null || isWidgetIncludingAllParentsVisible(parent)
+        val parent = widget.parentId?.let { id -> widgetsById[id] } ?: return true
+        return isWidgetIncludingAllParentsVisible(parent)
     }
 
     private fun getItemViewType(widget: Widget): Int {
```

I've checked in the bytecode output that it indeed currently doesn't do tail recursion optimization.